### PR TITLE
Add support for memberlist bind network configuration

### PR DESCRIPTION
**Pull request by rubenvp8510**
Fixes https://github.com/grafana/tempo-operator/issues/1060

### CODE REVIEW

File: internal/manifests/config/build.go [hunk 1]
```diff
@@ -256,3 +258,17 @@ func renderTempoQueryTemplate(opts tempoQueryOptions) ([]byte, error) {
...
 	return cfg, nil
 }
+
+func gossipRingInstanceAddr(spec v1alpha1.HashRingSpec) string {
+	var instanceAddr string
+	switch spec.MemberList.InstanceAddrType {
+	case v1alpha1.InstanceAddrPodIP:
+		instanceAddr = fmt.Sprintf("${%s}", memberlist.GossipInstanceAddrEnvVarName)
+	case v1alpha1.InstanceAddrDefault:
+		// Do nothing use loki defaults
```
Comment (andreasgerstmayr) [inline internal/manifests/config/build.go:268]:
```suggestion
		// Do nothing use Tempo defaults
```

File: internal/manifests/memberlist/gossip.go [hunk 2]
```diff
@@ -46,3 +51,55 @@ func BuildGossip(tempo v1alpha1.TempoStack) *corev1.Service {
 		},
 	}
 }
+
+func resetEnvVar(podSpec *corev1.PodSpec, name string) {
```
Comment (andreasgerstmayr) [inline internal/manifests/memberlist/gossip.go:55]: I'm not sure why we need this? We don't let users set any custom env vars, only the operator can set them. So afaics it's impossible that the env var is already set and needs to be unset.

**Review by andreasgerstmayr — COMMENTED** [review]
The `compactor`, `distributor` and `ingester` get configured with the new setting, what about the other components `querier` and `query-frontend`? Do they not need the memberlist?

I think even if they don't need it now, I'd still add it, in case they need memberlist in the future. Wdyt?

**Comment by rubenvp8510** [comment]
> The `compactor`, `distributor` and `ingester` get configured with the new setting, what about the other components `querier` and `query-frontend`? Do they not need the memberlist?
> 
> I think even if they don't need it now, I'd still add it, in case they need memberlist in the future. Wdyt?

Is not needed but don't hurt to put it on those. I'll do the change!

**Comment by rubenvp8510** [comment]
@IshwarKanse Can you help me to test this? Thanks

**Comment by IshwarKanse** [comment]
@rubenvp8510 yes, I'll test the fix today.

**Comment by IshwarKanse** [comment]
@rubenvp8510 I tried with this CR

```
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  name: simplest
spec:
  hashRing:
    memberlist:
      instanceAddrType: "podIP"
  storage:
    secret:
      name: minio
      type: s3
  storageSize: 200M
  resources:
    total:
      limits:
        memory: 2Gi
        cpu: 2000m
  template:
    queryFrontend:
      jaegerQuery:
        enabled: true
        ingress:
          type: ingress
```

The TempoStack pods are failing due to config error:
```
 % oc get pods
NAME                                             READY   STATUS             RESTARTS      AGE
minio-679d9f746b-nj678                           1/1     Running            0             12m
tempo-simplest-compactor-5475bd75b4-t5pjt        0/1     CrashLoopBackOff   4 (72s ago)   2m50s
tempo-simplest-distributor-777858f46c-2lvmv      0/1     CrashLoopBackOff   4 (71s ago)   2m50s
tempo-simplest-ingester-0                        0/1     CrashLoopBackOff   4 (74s ago)   2m50s
tempo-simplest-querier-6c47bbbbcc-v6pnx          0/1     CrashLoopBackOff   4 (81s ago)   2m50s
tempo-simplest-query-frontend-7fb999dfc8-sw7qk   2/3     CrashLoopBackOff   4 (66s ago)   2m50s

% oc logs tempo-simplest-ingester-0
failed parsing config: failed to parse configFile /conf/tempo.yaml: yaml: unmarshal errors:
  line 35: field instance_addr not found in type ring.Config

% oc extract cm/tempo-simplest --confirm
tempo.yaml
overrides.yaml
tempo-query-frontend.yaml
tempo-query.yaml

% cat tempo.yaml 
compactor:
  compaction:
    block_retention: 48h0m0s
  ring:
    kvstore:
      store: memberlist
distributor:
  receivers:
    jaeger:
      protocols:
        thrift_http:
          endpoint: 0.0.0.0:14268
        thrift_binary:
          endpoint: 0.0.0.0:6832
        thrift_compact:
          endpoint: 0.0.0.0:6831
        grpc:
          endpoint: 0.0.0.0:14250
    zipkin:
    otlp:
      protocols:
        grpc:
          endpoint: 0.0.0.0:4317
        http:
          endpoint: 0.0.0.0:4318
  ring:
    kvstore:
      store: memberlist
    instance_addr: ${HASH_RING_INSTANCE_ADDR}
ingester:
  lifecycler:
    ring:
      kvstore:
        store: memberlist
      instance_addr: ${HASH_RING_INSTANCE_ADDR}
      replication_factor: 1
    tokens_file_path: /var/tempo/tokens.json
  max_block_duration: 10m
memberlist:
  abort_if_cluster_join_fails: false
  join_members:
  - tempo-simplest-gossip-ring
multitenancy_enabled: false
querier:
  max_concurrent_queries: 20
  frontend_worker:
    frontend_address: tempo-simplest-query-frontend-discovery:9095
    grpc_client_config:
      tls_enabled: true
      tls_cert_path:  /var/run/tls/server/tls.crt
      tls_key_path: /var/run/tls/server/tls.key
      tls_ca_path: /var/run/ca/service-ca.crt
      tls_server_name: tempo-simplest-query-frontend.test.svc.cluster.local
      tls_min_version: VersionTLS13
  search:
    external_hedge_requests_at: 8s
    external_hedge_requests_up_to: 2
internal_server:
  enable: true
  http_listen_address: ""
  tls_min_version: VersionTLS13
  http_tls_config:
    cert_file: /var/run/tls/server/tls.crt
    key_file: /var/run/tls/server/tls.key
server:
  grpc_server_max_recv_msg_size: 4194304
  grpc_server_max_send_msg_size: 4194304
  http_listen_port: 3200
  http_server_read_timeout: 30s
  http_server_write_timeout: 30s
  log_format: logfmt
  tls_min_version: VersionTLS13
  grpc_tls_config:
    cert_file:  /var/run/tls/server/tls.crt
    key_file: /var/run/tls/server/tls.key
    client_ca_file: /var/run/ca/service-ca.crt
    client_auth_type: RequireAndVerifyClientCert
  http_tls_config:
    cert_file:  /var/run/tls/server/tls.crt
    client_auth_type: RequireAndVerifyClientCert
    key_file: /var/run/tls/server/tls.key
    client_ca_file: /var/run/ca/service-ca.crt
storage:
  trace:
    backend: s3
    blocklist_poll: 5m
    cache: none
    s3:
      endpoint: minio:9000
      bucket: tempo
      insecure: true
    local:
      path: /var/tempo/traces
    wal:
      path: /var/tempo/wal
usage_report:
  reporting_enabled: false
query_frontend:
  search:
    concurrent_jobs: 2000
    max_duration: 0s
    default_result_limit: 20
ingester_client:
  grpc_client_config:
    tls_enabled: true
    tls_cert_path:  /var/run/tls/server/tls.crt
    tls_key_path: /var/run/tls/server/tls.key
    tls_ca_path: /var/run/ca/service-ca.crt
    tls_server_name: tempo-simplest-ingester.test.svc.cluster.local
    tls_insecure_skip_verify: false
    tls_min_version: VersionTLS13
```

**Comment by IshwarKanse** [comment]
@rubenvp8510 The CR is not taking the instanceAddrType. Created TempoStack with.
```
  hashRing:
    memberlist:
      instanceAddrType: "podIP"
```
The instance created is missing the parameter. 
```
 % oc get tempostacks.tempo.grafana.com simplest -o yaml
apiVersion: tempo.grafana.com/v1alpha1
kind: TempoStack
metadata:
  creationTimestamp: "2024-10-18T04:21:18Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: tempo-operator
    tempo.grafana.com/distribution: openshift
  name: simplest
  namespace: test
  resourceVersion: "46067"
  uid: 906660b1-3d33-403a-a2ea-c1f401af8a84
spec:
  hashRing:
    memberlist: {}
  images: {}

```

**Review by andreasgerstmayr — APPROVED** [review]

**Pull request merged**